docs/source/MESSAGES.rst

```diff
@@ -1291,6 +1418,7 @@ ResetDistributed (Code 130)
 Server requests to reset our parent and children
 
 :Code: 127 (0x7F)
+:Status: UNKNOWN
```

Comment:
I've seen the server send this message to a Nicotine+ user twice, most recently last year, but I don't know what the conditions are.

Reply:
That's a good suggestion. Could also happen perhaps if a user gets banned.

I'll try it out as I'm trying to understand the distributed network a bit better but I'm currently a bit low on energy so it might take a while before I do

Reply:
If you have any specific questions about the distributed network, I could try to answer them. I spent quite a lot of time trying to understand it as well when working on the implementation in Nicotine+.

Reply:
Yeah perhaps some input would be good. My understanding about how it works has changed a bit over the past weeks, so there's a lot of speculation in the explanation.

*Root selection*

I initially thought the server would specifically select the distributed roots; when a client would get a search request from the server it would mean that that peer would become a root permanently and the server would then pass search requests it received to each of its branch roots. This does not seem to be the case (official clients do not stop looking for parents when they receive a search request from the server). Rather, it seems like the server collects all search requests for a second, then sends them out to the peers with the highest upload speed.

Example: the server received 80 search requests in a second, it takes the 4 peers with the highest upload speed and pushes 20 requests to each of them. If it received 90 then the server picks 5 peers.

There's some issues to this logic though, some might solve themselves naturally while others might not. Some I can think off right now:

* If the amount of search requests is low then there will only be 1 branch root with a lot of children. If he disconnects all subsequent search requests suddenly get redirected to the next peer with the highest upload speed. If he doesn't have children then basically the next few seconds search requests will not reach a large section of the network. -> This might resolve itself naturally if the entire branch does not break up immediately and the peers at level 1 can quickly reconnect with the 
* If a new peer connects with a higher speed all search requests will be redirected to him while he probably does not have any children yet which would again cause a short period where search requests do not reach a large section of the network
* The BranchLevel and BranchRoot messages sent to the server are effectively useless in this scenario

This was based on 4 observations:

* Peers only receive 20 search requests / second
* If your upload speed is high enough but you don't try to attach to a parent then the server will sporadically send search requests
* There really doesn't seem to be a lot of branch roots, sometimes maybe like 2-3?
* Client never seems to stop searching parents if he is branch root (ToggleParentSearch=false), not when he receives a server message nor when he gets his first child.

*Potential parent selection*

The selection of the potential parents which a peer receives every 60s seems random, the only 2 criteria looked at is the upload speed and whether the peer is accepting children. The speed of the potential parent needs to be greater than yours. One caveat: the upload speed seems cached (don't know for how long or when), so it possible that you receive a potential parent with a speed less than yours but it's a very small percentage.

Some potential issues:

* This could create bottlenecks if the parent has about the same speed
* It also allows a peer with very low speed to connect to a root for example -> might resolve itself naturally because the chances of a low speed peer connecting to a high speed one is low because there's more peers with a higher speed if yours is low

*Client / message specific*

Well, I still don't understand why roots (level = 0) just push the server messages straight to their children (level = 1) and only those "unpack" that message and in turn send it as distributed message to their children (level = 2).

The magic number 49 remains a mystery. 

And well, I'm still not 100% sure when to disconnect the children but could investigate:

* If server connection is lost?
* If parent connection is lost?

Reply:
> Well, I still don't understand why roots (level = 0) just push the server messages straight to their children (level = 1) and only those "unpack" that message and in turn send it as distributed message to their children (level = 2).

I don't know why branch roots currently behave like this, but I can imagine being able to transmit an embedded message would be useful new distributed messages were ever added in the future. A client that doesn't understand a newer message could just transmit the embedded message as-is, until it reaches a newer client that can unpack it.

I wonder if I should add some code to turn unknown distributed messages into embedded ones, and transmit them to child peers.

> The magic number 49 remains a mystery.

Yep, I haven't figured this one out.

> And well, I'm still not 100% sure when to disconnect the children but could investigate:
> If server connection is lost?

I know the official client doesn't do this, but Nicotine+ just close all peer connections when disconnecting from the server to make things easier.

> If parent connection is lost?

Would be good to verify what the official client does, but I just let the normal 60 second peer connection timeout take care of closing them.

Reply:
> I wonder if I should add some code to turn unknown distributed messages into embedded ones, and transmit them to child peers.

SoulseekQt doesn't do this, I just checked. If SoulseekQt receives an unknown distributed message, or an embedded message with an unknown code, they are not sent to its child peers.

Reply:
Regarding server message 130, it looks like the server sends it if you send the name of an invalid/offline user as your branch root.

Reply:
Thanks for checking, I documented it